Fix AssemblyName cache hash and key

sdmaclea opened the pull request:
Fixes #24135

* `AssemblyName::Hash(DWORD dwIncludeFlags)` was including should be ignored attributes in its hash.
* `AssemblyNative::Load` was using stackMark in a case where it wasn't needed.
* Compiler started erroring because `RuntimeAssembly.m_assembly` was unsused in C#.  It is used in native code.  Disable the warning.
* Add ContextualReflection `LoadWithPartialName` test case.
* Remove unnecessary `NoInlining` attribute
* Disable broken CoreFX test case dotnet/corefx#37071

jkotas commented on src/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs:
```diff
@@ -304,38 +304,11 @@ public override IList<CustomAttributeData> GetCustomAttributesData()
-            AssemblyName an = new AssemblyName();
-
-            an.Name = assemblyString;
-            an.nInit(out assemblyFromResolveEvent, true);
```
You can also delete both arguments from the `nInit` FCall. Nothing is using them after this change.

jkotas commented:
`src/System/Reflection/Assembly.CoreCLR.cs(40,20): error CS0246: The type or namespace name 'FileNotFoundException' could not be found (are you missing a using directive or an assembly reference?)`

jkotas commented:
Do you plan to add any tests for this?

sdmaclea commented:
> Do you plan to add any tests for this?

The `Assembly.LoadWithPartialName` test is completely missing from the ContextualReflection test. I will add it.

I expected the ContextualReflections test to catch the `Assembly.Load(string)` failure, but it didn't.  The corefx rewrite of the test did however. I need to look closer at it, to figure out why this use case was missed. If/when I can repro, I'll edit/add the test.

I am adding tests in corefx.  These found the issue, but since they are not yet passing they are disabled until these bits make it.

sdmaclea committed:
Remove m_assembly warning

sdmaclea committed:
Remove unnecessary MethodImplOptions.NoInlining

sdmaclea committed:
Add ContextualReflection LoadWithPartialName case

sdmaclea commented:
Need to diagnose the root cause further. This is not fixing the observed issue.

sdmaclea committed:
Fix AssemblyName hash function

sdmaclea committed:
AssemblyNative::Load fix stackMark usage

Do not use the stackMark if (ptrLoadContextBinder != NULL)

sdmaclea commented:
Spent a lot of time in debugger root causing this:

- The binder cache was not finding assemblies already loaded because the key lookup was failing.  
- The AssemblyName Hash function was using the public_key and public_token when it wasn't supposed to.  
- Also we were populating the parent assembly using the stack mark when we didn't need to. This led to cases where the wrong binder_id was used as the key.

I have added the LoadWithPartialName test.  The corefx test will catch the binder cache issue. I used that to debug the failure. I should be able to create a lighter test, but I'll put it in a separate PR.  

I would like this fix to make preview5.

sdmaclea commented:
There are two "regressions" in the test runs.
https://github.com/dotnet/corefx/blob/13cd96122c25328a1b180a6289923657180a7b69/src/System.Runtime/tests/System/ActivatorTests.netcoreapp.cs#L263-L265, Is testing a resolve event is fired when the assemblyName is illegal/malformed. That is the removed CreateAssemblyName function. I'll remove that change.

The other regression is. 
https://github.com/dotnet/corefx/blob/13cd96122c25328a1b180a6289923657180a7b69/src/System.Runtime.Loader/tests/DefaultContext/DefaultLoadContextTest.cs#L260-L270

The test is bad. It is regressing because the assembly cache is now finding the `System.Runtime` assembly which was loaded before clearing the `LoadedFromContext` flag. So this assert is failing. Moving the clear earlier in the function seems to resolve the failure.

sdmaclea commented:
This looks like it will miss preview5.

sdmaclea committed:
Temporarily disable DefaultContextOverrideTPA

Test is failing due to a logic error.
Fix is pending in https://github.com/dotnet/corefx/pull/37071

jkotas commented:
> resolve event is fired when the assemblyName is illegal/malformed

FWIW, I would not mind losing this "feature". Illegal/malformed assembly names will fail everywhere else, except this one place that tries to recover from them. It can be done separately / later.

jkotas reviewed (approved):
Thanks!
